cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreServerModule.java

```diff
@@ -0,0 +1,113 @@
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.conf.SConfiguration;
+import co.cask.cdap.security.store.DefaultSecureStoreService;
+import co.cask.cdap.security.store.DummySecureStoreService;
+import co.cask.cdap.security.store.FileSecureStoreService;
+import co.cask.cdap.security.store.SecureStoreService;
+import co.cask.cdap.security.store.SecureStoreUtils;
+import co.cask.cdap.security.store.secretmanager.SecretManagerSecureStoreService;
+import com.google.common.base.Strings;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Module;
+import com.google.inject.PrivateModule;
+import com.google.inject.Provider;
+import com.google.inject.Scopes;
+import com.google.inject.Singleton;
+import com.google.inject.name.Names;
+
+/**
+ * Server side guice bindings for security store related classes.
+ */
+public class SecureStoreServerModule {
+  public static final String DELEGATE_SECURE_STORE_SERVICE = "delegateSecureStoreService";
+
+  public Module getSecureStoreModule() {
+    return new PrivateModule() {
+      @Override
+      protected void configure() {
+        bind(SecureStoreService.class)
+          .annotatedWith(Names.named(DELEGATE_SECURE_STORE_SERVICE))
+          .toProvider(SecureStoreServiceProvider.class);
+
+        bind(DefaultSecureStoreService.class).in(Scopes.SINGLETON);
+        bind(SecureStore.class).to(DefaultSecureStoreService.class);
+        bind(SecureStoreManager.class).to(DefaultSecureStoreService.class);
+        bind(SecureStoreService.class).to(DefaultSecureStoreService.class);
+        expose(SecureStore.class);
+        expose(SecureStoreManager.class);
+        expose(SecureStoreService.class);
+      }
+    };
+  }
+
+  /**
+   * Secure store service provider.
+   *
+   * file binds to FileSecureStoreService
+   * kms  binds to KMSSecureStoreService
+   * ext  binds to SecretManagerSecureStoreService
+   * none binds to DummySecureStoreService
+   *
+   * TODO: simplify this bindings after CDAP-14668 is fixed.
+   */
+  @Singleton
+  private static final class SecureStoreServiceProvider implements Provider<SecureStoreService> {
+    private final CConfiguration cConf;
+    private final SConfiguration sConf;
+    private final Injector injector;
+
+    @Inject
+    private SecureStoreServiceProvider(final CConfiguration cConf, SConfiguration sConf, Injector injector) {
+      this.cConf = cConf;
+      this.sConf = sConf;
+      this.injector = injector;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public SecureStoreService get() {
+      if (SecureStoreUtils.isFileBacked(cConf) &&
+        !Strings.isNullOrEmpty(sConf.get(Constants.Security.Store.FILE_PASSWORD))) {
+        return injector.getInstance(FileSecureStoreService.class);
+      } else if (SecureStoreUtils.isFileBacked(cConf)) {
+        throw new IllegalArgumentException("Only KMS based provider is supported in distributed mode. " +
```

Comment:
I don't understand the logic. If there is no `FILE_PASSWORD` set, then it means it is distributed env?